File: Lab_7/files!/2/Semaphore1.c
#include <stdio.h>
#include <pthread.h>
#include <semaphore.h>

#define NUM_ITER 1000

int counter = 0;//shared memory
sem_t s; //binary semaphore

void* increment(void* arg)
{
    for(int i=0;i<NUM_ITER;i++) 
	{
        sem_wait(&s); //enter critical section -> take a key
        counter++;//perform update
        sem_post(&s); //leave critical section -> return key
    }
    return NULL;
}
int main()
{
    pthread_t t1, t2;
    sem_init(&s, 0, 1); // binary semaphore initialized to 1

    pthread_create(&t1, NULL, increment, NULL);
    pthread_create(&t2, NULL, increment, NULL);

    pthread_join(t1, NULL);
    pthread_join(t2, NULL);

    printf("Final counter: %d\n", counter);
    sem_destroy(&s);
    return 0;
}

File: Lab_7/files!/2/Semaphore1.java
import java.util.concurrent.Semaphore;
//https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Semaphore.html
public class Semaphore1
{
    static int counter=0;//shared memory -> possible race condition
    static Semaphore s=new Semaphore(1);//create a binary semaphore
    public static void main(String[] args) throws InterruptedException
	{
        Runnable task = () -> {
            for (int i = 0; i < 1000; i++) {
                try {
                    s.acquire();//enter critical section -> take a key
                    counter++;//perform update
                } catch (InterruptedException e) {}
                finally {
                    s.release();//leave critical section -> return key
                }
            }
        };

        Thread t1 = new Thread(task);
        Thread t2 = new Thread(task);

        t1.start();
        t2.start();
        t1.join();//join back to main thread before final print happens
        t2.join();

        System.out.println("Final counter: " + counter);
    }
}